In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("All libraries imported successfully!")

All libraries imported successfully!


In [6]:
import os
print(os.getcwd())

c:\Users\HP\climate-challenge-week0\src\notebooks


In [8]:
df = pd.read_csv(r"C:\Users\HP\climate-challenge-week0\ethiopia.csv")
df.head()

,YEAR,DOY,T2M,T2M_MAX,T2M_MIN,T2M_RANGE,PRECTOTCORR,RH2M,WS2M,WS2M_MAX,PS,QV2M
0,2015,1,11.73,22.75,3.44,19.31,0.0,41.79,2.73,5.07,77.13,4.00
1,2015,2,12.30,24.01,4.09,19.92,0.0,33.29,2.39,4.19,77.14,3.35
2,2015,3,12.49,24.17,3.97,20.20,0.0,33.83,1.77,2.76,77.11,3.43
3,2015,4,14.08,23.78,6.90,16.88,0.0,38.84,0.87,1.28,77.07,4.60
4,2015,5,14.06,23.15,7.32,15.83,0.0,47.07,1.34,2.14,77.01,5.58


In [9]:
df['Country'] = "Ethiopia"
df.head()

,YEAR,DOY,T2M,T2M_MAX,T2M_MIN,T2M_RANGE,PRECTOTCORR,RH2M,WS2M,WS2M_MAX,PS,QV2M,Country
0,2015,1,11.73,22.75,3.44,19.31,0.0,41.79,2.73,5.07,77.13,4.00,Ethiopia
1,2015,2,12.30,24.01,4.09,19.92,0.0,33.29,2.39,4.19,77.14,3.35,Ethiopia
2,2015,3,12.49,24.17,3.97,20.20,0.0,33.83,1.77,2.76,77.11,3.43,Ethiopia
3,2015,4,14.08,23.78,6.90,16.88,0.0,38.84,0.87,1.28,77.07,4.60,Ethiopia
4,2015,5,14.06,23.15,7.32,15.83,0.0,47.07,1.34,2.14,77.01,5.58,Ethiopia


In [10]:
df['Date'] = pd.to_datetime(df['YEAR'] * 1000 + df['DOY'], format="%Y%j")
df.head()

,YEAR,DOY,T2M,T2M_MAX,T2M_MIN,T2M_RANGE,PRECTOTCORR,RH2M,WS2M,WS2M_MAX,PS,QV2M,Country,Date
0,2015,1,11.73,22.75,3.44,19.31,0.0,41.79,2.73,5.07,77.13,4.00,Ethiopia,2015-01-01
1,2015,2,12.30,24.01,4.09,19.92,0.0,33.29,2.39,4.19,77.14,3.35,Ethiopia,2015-01-02
2,2015,3,12.49,24.17,3.97,20.20,0.0,33.83,1.77,2.76,77.11,3.43,Ethiopia,2015-01-03
3,2015,4,14.08,23.78,6.90,16.88,0.0,38.84,0.87,1.28,77.07,4.60,Ethiopia,2015-01-04
4,2015,5,14.06,23.15,7.32,15.83,0.0,47.07,1.34,2.14,77.01,5.58,Ethiopia,2015-01-05


In [11]:
df['Month'] = df['Date'].dt.month
df.head()

,YEAR,DOY,T2M,T2M_MAX,T2M_MIN,T2M_RANGE,PRECTOTCORR,RH2M,WS2M,WS2M_MAX,PS,QV2M,Country,Date,Month
0,2015,1,11.73,22.75,3.44,19.31,0.0,41.79,2.73,5.07,77.13,4.00,Ethiopia,2015-01-01,1
1,2015,2,12.30,24.01,4.09,19.92,0.0,33.29,2.39,4.19,77.14,3.35,Ethiopia,2015-01-02,1
2,2015,3,12.49,24.17,3.97,20.20,0.0,33.83,1.77,2.76,77.11,3.43,Ethiopia,2015-01-03,1
3,2015,4,14.08,23.78,6.90,16.88,0.0,38.84,0.87,1.28,77.07,4.60,Ethiopia,2015-01-04,1
4,2015,5,14.06,23.15,7.32,15.83,0.0,47.07,1.34,2.14,77.01,5.58,Ethiopia,2015-01-05,1


In [12]:
import numpy as np
df.replace(-999, np.nan, inplace=True)
print("Missing values per column:")
print(df.isna().sum())

Missing values per column:
YEAR           0
DOY            0
T2M            0
T2M_MAX        0
T2M_MIN        0
T2M_RANGE      0
PRECTOTCORR    0
RH2M           0
WS2M           0
WS2M_MAX       0
PS             0
QV2M           0
Country        0
Date           0
Month          0
dtype: int64


In [13]:
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 0


In [14]:
df.describe()

,YEAR,DOY,T2M,T2M_MAX,T2M_MIN,T2M_RANGE,PRECTOTCORR,RH2M,WS2M,WS2M_MAX,PS,QV2M,Date,Month
count,4108.000000,4108.000000,4108.00000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108,4108.000000
mean,2020.131451,180.121227,16.06850,23.199175,10.227544,12.971631,3.633795,68.408588,1.979998,3.575246,77.037529,9.697724,2020-08-15 12:00:00,6.423564
min,2015.000000,1.000000,10.03000,15.650000,1.170000,3.560000,0.000000,14.420000,0.460000,0.790000,76.360000,2.120000,2015-01-01 00:00:00,1.000000
25%,2017.000000,86.000000,14.82000,21.110000,8.417500,9.830000,0.020000,59.030000,1.407500,2.720000,76.980000,8.070000,2017-10-23 18:00:00,3.000000
50%,2020.000000,179.000000,16.04000,22.740000,10.990000,13.030000,0.820000,71.120000,1.970000,3.530000,77.040000,10.270000,2020-08-15 12:00:00,6.000000
75%,2023.000000,272.000000,17.36000,25.170000,12.220000,16.020000,4.580000,80.802500,2.480000,4.370000,77.100000,11.710000,2023-06-08 06:00:00,9.000000
max,2026.000000,366.000000,21.53000,30.930000,15.680000,23.240000,82.300000,91.930000,4.130000,7.490000,77.370000,13.520000,2026-03-31 00:00:00,12.000000
std,3.248907,106.294767,1.89805,2.751471,2.607256,3.821239,6.289061,14.735838,0.689093,1.123721,0.099484,2.362462,NaN,3.477046


In [15]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4108 entries, 0 to 4107
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   YEAR         4108 non-null   int64         
 1   DOY          4108 non-null   int64         
 2   T2M          4108 non-null   float64       
 3   T2M_MAX      4108 non-null   float64       
 4   T2M_MIN      4108 non-null   float64       
 5   T2M_RANGE    4108 non-null   float64       
 6   PRECTOTCORR  4108 non-null   float64       
 7   RH2M         4108 non-null   float64       
 8   WS2M         4108 non-null   float64       
 9   WS2M_MAX     4108 non-null   float64       
 10  PS           4108 non-null   float64       
 11  QV2M         4108 non-null   float64       
 12  Country      4108 non-null   str           
 13  Date         4108 non-null   datetime64[us]
 14  Month        4108 non-null   int32         
dtypes: datetime64[us](1), float64(10), int32(1), int64(2), str(1)
memo

In [16]:
df.columns.tolist()

['YEAR',
 'DOY',
 'T2M',
 'T2M_MAX',
 'T2M_MIN',
 'T2M_RANGE',
 'PRECTOTCORR',
 'RH2M',
 'WS2M',
 'WS2M_MAX',
 'PS',
 'QV2M',
 'Country',
 'Date',
 'Month']

In [17]:
import numpy as np
df.replace(-999, np.nan, inplace=True)
print("Missing values per column:")
print(df.isna().sum())

Missing values per column:
YEAR           0
DOY            0
T2M            0
T2M_MAX        0
T2M_MIN        0
T2M_RANGE      0
PRECTOTCORR    0
RH2M           0
WS2M           0
WS2M_MAX       0
PS             0
QV2M           0
Country        0
Date           0
Month          0
dtype: int64


In [18]:
missing_percent = (df.isna().sum() / len(df)) * 100
print("Missing value percentages:")
print(missing_percent[missing_percent > 0])

Missing value percentages:
Series([], dtype: float64)


In [19]:
print(f"Duplicate rows found: {df.duplicated().sum()}")
df = df.drop_duplicates()
print(f"Shape after removing duplicates: {df.shape}")

Duplicate rows found: 0
Shape after removing duplicates: (4108, 15)


In [20]:
df.describe()

,YEAR,DOY,T2M,T2M_MAX,T2M_MIN,T2M_RANGE,PRECTOTCORR,RH2M,WS2M,WS2M_MAX,PS,QV2M,Date,Month
count,4108.000000,4108.000000,4108.00000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108.000000,4108,4108.000000
mean,2020.131451,180.121227,16.06850,23.199175,10.227544,12.971631,3.633795,68.408588,1.979998,3.575246,77.037529,9.697724,2020-08-15 12:00:00,6.423564
min,2015.000000,1.000000,10.03000,15.650000,1.170000,3.560000,0.000000,14.420000,0.460000,0.790000,76.360000,2.120000,2015-01-01 00:00:00,1.000000
25%,2017.000000,86.000000,14.82000,21.110000,8.417500,9.830000,0.020000,59.030000,1.407500,2.720000,76.980000,8.070000,2017-10-23 18:00:00,3.000000
50%,2020.000000,179.000000,16.04000,22.740000,10.990000,13.030000,0.820000,71.120000,1.970000,3.530000,77.040000,10.270000,2020-08-15 12:00:00,6.000000
75%,2023.000000,272.000000,17.36000,25.170000,12.220000,16.020000,4.580000,80.802500,2.480000,4.370000,77.100000,11.710000,2023-06-08 06:00:00,9.000000
max,2026.000000,366.000000,21.53000,30.930000,15.680000,23.240000,82.300000,91.930000,4.130000,7.490000,77.370000,13.520000,2026-03-31 00:00:00,12.000000
std,3.248907,106.294767,1.89805,2.751471,2.607256,3.821239,6.289061,14.735838,0.689093,1.123721,0.099484,2.362462,NaN,3.477046


## Interpretation of Summary Statistics - Ethiopia Climate Data (2015-2026)

### Temperature Analysis (T2M - Mean Daily Temperature)

- **Average Temperature:** 16.07°C over the entire period
- **Temperature Range:** 10.03°C (minimum) to 21.53°C (maximum)
- **Standard Deviation:** 1.90°C, indicating moderate seasonal variation
- **Median vs Mean:** 16.04°C vs 16.07°C - nearly identical, suggesting symmetrical temperature distribution

### Maximum Temperature (T2M_MAX)

- **Average Daily Max:** 23.20°C
- **Extreme Heat Days:** Maximum recorded at 30.93°C
- **Range:** 15.65°C to 30.93°C (15.28°C spread)

### Minimum Temperature (T2M_MIN)

- **Average Daily Min:** 10.23°C
- **Cool Nights:** Minimum recorded at 1.17°C
- **Range:** 1.17°C to 15.68°C

### Daily Temperature Range (T2M_RANGE)

- **Average Daily Range:** 12.97°C
- **Range Spread:** 3.56°C to 23.24°C
- **Implication:** Large diurnal temperature variation typical of highland climates (Ethiopian highlands)

### Precipitation Analysis (PRECTOTCORR)

- **Average Daily Rainfall:** 3.63 mm/day
- **Rainfall Pattern:** Highly skewed - 75% of days have ≤4.58 mm, but maximum reaches 82.30 mm
- **Zero Rainfall Days:** Minimum of 0.00 mm indicates many dry days
- **Standard Deviation:** 6.29 mm (higher than mean, confirming skewness toward dry conditions)

### Humidity (RH2M)

- **Average Relative Humidity:** 68.41%
- **Range:** 14.42% to 91.93%
- **Interpretation:** Generally humid conditions, typical of tropical climate

### Wind Speed (WS2M)

- **Average Wind Speed:** 1.98 m/s (light breeze)
- **Maximum Gust:** 7.49 m/s (moderate breeze)
- **Generally calm conditions** with occasional stronger winds

### Atmospheric Pressure (PS)

- **Average Pressure:** 77.04 kPa
- **Very stable** (standard deviation only 0.10 kPa)
- **Elevation indicator:** Lower than sea-level pressure (101.3 kPa), confirming Ethiopia's highland topography

### Key Observations for COP32 Negotiations:

1. **Temperature Trend:** Ethiopia shows moderate temperatures (avg 16°C) but with extremes reaching 31°C, indicating warming pressures

2. **Precipitation Vulnerability:** Rainfall is highly variable (0-82 mm/day), with many dry days - suggests drought risk and water insecurity

3. **Diurnal Range:** Large day-night temperature swings (avg 13°C) affects agriculture and energy demand

4. **Data Quality:** No obvious missing value issues with 4,108 complete observations from 2015-2026

In [1]:
from scipy import stats

# Variables to check for outliers
variables = ['T2M', 'T2M_MAX', 'T2M_MIN', 'PRECTOTCORR', 'RH2M', 'WS2M', 'WS2M_MAX']

# Store outlier counts
outlier_counts = {}

for var in variables:
    # Drop NaN values for Z-score calculation
    clean_data = df[var].dropna()
    z_scores = np.abs(stats.zscore(clean_data))
    outlier_counts[var] = (z_scores > 3).sum()
    print(f"{var}: {outlier_counts[var]} outliers | {outlier_counts[var]/len(clean_data)*100:.2f}% of data")

NameError: name 'df' is not defined

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

df = pd.read_csv(r"C:\Users\HP\climate-challenge-week0\ethiopia.csv")

df['Country'] = "Ethiopia"
df['Date'] = pd.to_datetime(df['YEAR'] * 1000 + df['DOY'], format="%Y%j")
df['Month'] = df['Date'].dt.month

# Replace -999 with NaN
df.replace(-999, np.nan, inplace=True)

# Outlier detection (no scipy needed)
variables = ['T2M', 'T2M_MAX', 'T2M_MIN', 'PRECTOTCORR', 'RH2M', 'WS2M', 'WS2M_MAX']
for var in variables:
    clean_data = df[var].dropna()
    z_scores = np.abs((clean_data - clean_data.mean()) / clean_data.std())
    outliers = (z_scores > 3).sum()
    print(f"{var}: {outliers} outliers ({outliers/len(clean_data)*100:.2f}%)")

# Fill missing values
weather_cols = ['T2M', 'T2M_MAX', 'T2M_MIN', 'T2M_RANGE', 'PRECTOTCORR', 'RH2M', 'WS2M', 'WS2M_MAX', 'PS', 'QV2M']
df[weather_cols] = df[weather_cols].ffill()

# Save cleaned data
os.makedirs("../data", exist_ok=True)
df.to_csv("../data/ethiopia_clean.csv", index=False)

print("\n DONE! Cleaned data saved to ../data/ethiopia_clean.csv")
print(f" Shape: {df.shape}")
print(f" Date range: {df['Date'].min()} to {df['Date'].max()}")

T2M: 3 outliers (0.07%)
T2M_MAX: 0 outliers (0.00%)
T2M_MIN: 18 outliers (0.44%)
PRECTOTCORR: 95 outliers (2.31%)
RH2M: 13 outliers (0.32%)
WS2M: 3 outliers (0.07%)
WS2M_MAX: 5 outliers (0.12%)

 DONE! Cleaned data saved to ../data/ethiopia_clean.csv
 Shape: (4108, 15)
 Date range: 2015-01-01 00:00:00 to 2026-03-31 00:00:00


## Outlier Treatment Decision - Ethiopia

### Outlier Count Summary:
| Variable | Outliers (|Z|>3) | Percentage |
|----------|-----------------|------------|
| T2M      | 3               | 0.07% |
| T2M_MAX  | 0               | 0.00% |
| T2M_MIN  | 18              | 0.44% |
| PRECTOTCORR | 95           | 2.31% |
| RH2M     | 13              | 0.32% |
| WS2M     | 3               | 0.07% |
| WS2M_MAX | 5               | 0.12% |

### Decision: RETAIN all outliers

**Reasoning:**
1. **Climate data is naturally extreme** - Heat waves, cold snaps, and heavy rainfall are real phenomena, not errors
2. **Very low percentages** - All outliers represent <2.5% of total observations
3. **COP32 relevance** - Extreme events are critical for vulnerability assessment
4. **Values are plausible** - e.g., 30°C is realistic for Ethiopia

### Missing Value Treatment:
- **Method used:** Forward-fill (ffill) for weather variables
- **Reason:** Weather patterns are temporally correlated; yesterday's values are good proxies
- **Rows dropped:** None (no rows had >30% missing values)

### Export:
- Cleaned data saved to: `data/ethiopia_clean.csv`
- CSV excluded from Git (added to .gitignore)